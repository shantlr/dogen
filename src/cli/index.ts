import { Command } from 'commander';
import { generateDockerfile } from '../lib';
import { isFileExists } from '../lib/utils';
import { parseDogenConfigFile } from '../lib/dogen-config';
import { DogenConfig } from '../lib/types';

const cli = async (argv = process.argv) => {
  const prog = new Command();

  prog
    .command('gen [config-path]', {
      isDefault: true,
    })
    .description(`generate Dockerfile (default command)`)
    .option(
      '-w, --wizard',
      'Start with wizard, this is activated by default when no configuration is provided or detected'
    )
    .action(async (configPath?: string) => {
      try {
        if (configPath && !(await isFileExists(configPath))) {
          console.error(`config does not exists at '${configPath}'`);
          process.exit(1);
        }

        let config: DogenConfig = undefined;
        //#region Parse provided config
        if (configPath) {
          const parsed = await parseDogenConfigFile(configPath);
          config = parsed.config;
          if (parsed.warnings.length) {
            console.warn(parsed.warnings.map((c) => `WARN: ${c}`).join('\n'));
          }
        }
        //#endregion

        const { operation, dockerfilePath } = await generateDockerfile({
          dockerfileConflict: 'append',
          config,
        });
        console.log(`Dockerfile ${operation} at ${dockerfilePath}`);
      } catch (err) {
        switch (err?.message) {
          case 'DOCKERFILE_ALREADY_EXISTS': {
            console.error(`ERROR: Dockerfile already exists`);
            return;
          }
          default:
        }
        console.log(`Could not generated Dockerfile\nERROR:`, err);
      }
    });

  await prog.parseAsync(argv);
};

export default cli;
