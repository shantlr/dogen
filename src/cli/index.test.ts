import { readFile } from 'fs/promises';

import { DirectoryJSON, vol } from 'memfs';

import cli from '.';

vi.mock('fs/promises');

const setupVol = (json: DirectoryJSON, cwd: string) => {
  vol.fromJSON(json, cwd);
  const spy = vi.spyOn(process, 'cwd');
  spy.mockReturnValue(cwd);
};

describe('cli', () => {
  beforeEach(() => {
    vol.reset();
    vi.restoreAllMocks();
  });

  it('should create dockerfile', async () => {
    setupVol(
      {
        '/app/package.json': JSON.stringify({
          name: 'app',
          version: '1.0.0',
        }),
        '/app/yarn.lock': '',
      },
      '/app',
    );

    await cli(['ts-node', 'dogen']);
    expect((await readFile('./Dockerfile')).toString()).toMatchSnapshot();
  });

  it('should exit if no preset matched', async () => {
    setupVol({}, '/app');

    const spy = vi.spyOn(process, 'exit');
    try {
      await cli(['ts-node', 'dogen']);
    } catch {
      //
    }
    expect(spy).toHaveBeenCalledWith(1);
  });

  describe('presets', () => {
    describe('node-service', () => {
      it('should generate dockerfile', async () => {
        const packageJson = {
          name: 'express-server',
          version: '1.0.0',
          main: 'build/index.js',
        };
        setupVol(
          {
            '/service/package.json': JSON.stringify(packageJson),
            '/service/yarn.lock': '',
          },
          '/service',
        );

        await cli(['ts-node', 'dogen']);

        const dockerfile = (await readFile('./Dockerfile')).toString();

        expect(dockerfile).toMatchInlineSnapshot(`
          "#>>dogen
          FROM alpine:3.12 AS jq
          RUN apk add --update --no-cache jq

          FROM node:22-alpine AS base_node

          # Extract minimal fields for dependencies installation
          # This step avoid reinstalling node_modules due to field that is unrelated
          FROM jq AS extract_deps
          COPY ../../service/package.json /tmp/package.json
          RUN jq '{name, dependencies, devDependencies}' < /tmp/package.json > /tmp/deps.json

          # Install node_modules
          FROM base_node AS install_modules
          WORKDIR /app
          COPY --from=extract_deps /tmp/deps.json package.json
          COPY /service/yarn.lock yarn.lock
          RUN ["/bin/sh","-c","yarn","install","--pure-lockfile","--non-interactive","--cache-folder","./.ycache","&&","rm","-rf","./.ycache"]

          # Build
          FROM install_modules AS build
          WORKDIR /app
          COPY /service/package.json package.json
          RUN ["yarn","run","build"]

          FROM build AS service
          CMD ["yarn","run","start"]
          #<<dogen
          "
        `);
      });
    });
  });
});
