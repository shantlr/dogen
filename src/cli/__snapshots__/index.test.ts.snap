// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cli > should create dockerfile 1`] = `
"#>>dogen
FROM alpine:3.12 AS jq
RUN apk add --update --no-cache jq

FROM node:22-alpine AS base_node

# Extract minimal fields for dependencies installation
# This step avoid reinstalling node_modules due to field that is unrelated
FROM jq AS extract_deps
COPY ../../app/package.json /tmp/package.json
RUN jq '{name, dependencies, devDependencies}' < /tmp/package.json > /tmp/deps.json

# Install node_modules
FROM base_node AS install_modules
WORKDIR /app
COPY --from=extract_deps /tmp/deps.json package.json
COPY /app/yarn.lock yarn.lock
RUN ["/bin/sh","-c","yarn","install","--pure-lockfile","--non-interactive","--cache-folder","./.ycache","&&","rm","-rf","./.ycache"]

# Build
FROM install_modules AS build
WORKDIR /app
COPY /app/package.json package.json
RUN ["yarn","run","build"]

FROM build AS service
CMD ["yarn","run","start"]
#<<dogen
"
`;
